.github/CHANGELOG.md

```diff
@@ -2,6 +2,47 @@
 
 <h3>New features since last release</h3>
 
+* A transform `replace` has been added that allows you to define your own
+  custom decompositions or implementations of gates "on the fly"
+  without modifying the underlying code base.
+  [(#1482)](https://github.com/PennyLaneAI/pennylane/pull/1482)
+
+  For example, suppose you wanted to decompose all your Hadamard gates
+  as `[RY(np.pi/2), X]` rather than the built-in `[RZ(np.pi/2), RX(np.pi/2), RZ(np.pi/2)]`.
+  We can write a new decomposition function that accept a `wires` argument
+  (and an optional parameter arguments for parametrized gates), and
+  returns a custom list of operations in the same manner as the `decomposition` method
+  of a PennyLane `Operation`.
+
+  ```python
+  def custom_hadamard(wires):
+      return [qml.RY(np.pi/2, wires=wires), qml.PauliX(wires=wires)]
+  ```
+
+  Given a quantum function
+
+  ```python
+  def qfunc(x):
+      qml.Hadamard(wires=0)
+      qml.CNOT(wires=[0, 1])
+      qml.RX(x, wires=1)
+      qml.Hadamard(wires=1)
+      return qml.expval(qml.PauliZ(wires=1))
+  ```
```

Comment:
Another thought - the current `replace` transform will replace user-specified operations in the qfunc, but won't be re-applied later, if the device calls `op.decomposition()`, and the op decomposition contains an operation that was supposed to be replaced!

Reply:
> So this is more of a 'find-and-replace' operation applied to user-constructed qfuncs, rather than allowing users to define custom (fallback) decompositions?

I'm glad you brought this up, this was my intention and I hadn't thought of this other use case.

> I think this is possible to do; the transform simply loops through the tape operations, finds a matching operation, and replaces the op.decomposition() method.

What do you mean by "replace" the `op.decomposition()` method, like actually adjust the class method? Is that possible to do once, and then have it propagate to future instances of the operation?

> Another thought - the current replace transform will replace user-specified operations in the qfunc, but won't be re-applied later, if the device calls op.decomposition()

Yes this is true - okay then maybe replacing the decomposition is the better way of doing this.

Reply:
No worries!

> I'm glad you brought this up, 'find-and-replace' was my intention and I hadn't thought of this other use case.

If find-and-replace is the main use-case, I'm curious why you wouldn't just re-write your qfunc to make the replacements manually? This would be faster and have more compact code

> What do you mean by "replace" the op.decomposition() method, like actually adjust the class method? Is that possible to do once, and then have it propagate to future instances of the operation?

Yep! 

You would use something like

```python
with mock.patch("qml.ops.qubits.RX", "decomposition", new=custom_decomposition):
    # any decompositions or tape expansions done within the patched context
    # will use the users default decomposition
```

This ensures that the custom decomposition is only temporary, and doesn't affect other QNodes/qfuncs. See [`unittest.mock`](https://docs.python.org/3/library/unittest.mock.html) for more info.

Reply:
>  If find-and-replace of user-specified operations is the main use-case, I'm curious why you wouldn't just re-write your qfunc to make the replacements manually? This would be faster and have more compact code.
> I must admit, from the PR description, I originally thought that the transform was more 'specify custom fallback decompositions', which is much less accessible in PL.

You definitely could - but if you have more than one decomposition for an operation and want to test different ones, this would allow you just write the different decompositions and apply the transform with a different argument to the same underlying circuit. 

Definitely not averse to instead designing a mechanism for fallback decompositions though. Would this still be a transform?

Reply:
For what it's worth: I also thought that the PR addresses the fallback decompositions, and as Josh says, the other use case could be much more elegantly handled via direct control of the qfunc?

> You definitely could - but if you have more than one decomposition for an operation and want to test different ones, this would allow you just write the different decompositions and apply the transform with a different argument to the same underlying circuit.

I wonder if you could just do this via making the decomposition a qfunc argument:

```pytest
@qml.qnode(dev)
def qfunc(params, custom_hadamard):
    custom_hadamard(0)
    ...
```